packages/axelar-wasm-std/src/voting.rs

```diff
@@ -76,16 +160,201 @@ impl WeightedPoll {
             poll_size,
             votes: vec![Uint256::zero(); poll_size],
             status: PollStatus::InProgress,
+            voted: HashSet::new(),
         }
     }
 }
 
 impl Poll for WeightedPoll {
-    fn tally() -> Result<PollResult, Error> {
-        todo!()
+    fn tally(&mut self, block_height: u64) -> Result<PollResult, Error> {
+        if block_height < self.expires_at && self.voted.len() != self.snapshot.participants.len() {
+            return Err(Error::PollNotEnded {});
+        }
+
+        if self.status == PollStatus::Finished {
+            return Err(Error::PollNotInProgress {});
+        }
+
+        self.status = PollStatus::Finished;
+
+        Ok(PollResult {
+            poll_id: self.poll_id,
+            results: self
+                .votes
+                .iter()
+                .map(|v| *v >= self.snapshot.quorum.into())
+                .collect(),
+        })
     }
 
-    fn cast_vote(_sender: Addr, _poll_id: PollID, _votes: Vec<bool>) -> Result<PollStatus, Error> {
-        todo!()
+    fn cast_vote(
+        &mut self,
+        block_height: u64,
+        sender: &Addr,
+        votes: &[bool],
+    ) -> Result<PollStatus, Error> {
+        if !self.snapshot.is_participant(sender) {
+            return Err(Error::NotParticipant {});
+        }
+
+        if block_height > self.expires_at {
+            return Err(Error::PollExpired {});
+        }
+
+        if votes.len() != self.poll_size {
+            return Err(Error::InvalidVoteSize {});
+        }
+
+        if self.voted.contains(sender) {
+            return Err(Error::AlreadyVoted {});
+        }
+
+        if self.status != PollStatus::InProgress {
+            return Err(Error::PollNotInProgress {});
+        }
+
+        self.voted.insert(sender.clone());
+
+        for (tally, vote) in self.votes.iter_mut().zip(votes.iter()) {
+            if *vote {
+                *tally += self.snapshot.get_participant_weight(sender).unwrap();
```

Comment:
Just a thought, since we are looping through the tally for all of the messages here, we can calculate whether each message has achieved quorum and determine whether the poll is finished. Would probably be useful to do that and at least emit such in an event.